### ISSUE TITLE
Perhaps ty should not worry about parameter names to dunder methods like __getitem__()

### ISSUE THREAD

**Issue by mgedmin** [astral-sh/ty#4245]
### Summary

(I suspect this is going to be one of those "ty is technically correct, but maybe it should be more lenient" issues.)

I've a [Python script](https://github.com/mgedmin/scripts/blob/3e53b360b8cd94cad630a1c768e82f12607cb24f/list-repo-packages) that uses [python-debian](https://pypi.org/project/python-debian/) to analyze a set of packages available from an APT repository.  It has a Package class like this:

```python
from debian import deb822, debian_support

class Package:
    _pkg: deb822.Packages

    @property
    def version(self) -> debian_support.Version:
        return self._pkg.get_version()
```

This delegates to [debian.deb822](https://salsa.debian.org/python-debian-team/python-debian/-/blob/master/src/debian/deb822.py?ref_type=heads).Packages.get_version(), which is defined like this:

```python
class _VersionAccessorMixin:
    """Give access to Version keys as debian_support.Version objects."""
    def get_version(self: _HasVersionFieldProtocol) -> debian.debian_support.Version:
        return debian.debian_support.Version(self['Version'])
...
class Packages(Deb822, _PkgRelationMixin, _VersionAccessorMixin):
    ...
```

where `_HasVersionFieldProtocol` is defined as

```python
if TYPE_CHECKING:
    class _HasVersionFieldProtocol(Protocol):

        def __getitem__(self, s: str) -> str:
            pass
```

And ty doesn't like the overall result:

```console
$ uvx --with python-debian ty check list-repo-packages
error[invalid-argument-type]: Argument to bound method `_VersionAccessorMixin.get_version` is incorrect
   --> list-repo-packages:400:16
    |
400 |         return self._pkg.get_version()
    |                ^^^^^^^^^^^^^^^^^^^^^^^ Expected `_HasVersionFieldProtocol`, found `Packages`
info: type `Packages` is not assignable to protocol `_HasVersionFieldProtocol`
info: └── protocol member `__getitem__` is incompatible
info:     └── the parameter named `key` does not match `s` (and can be used as a keyword parameter)
info: Method defined here
    --> /home/mgedmin/.cache/uv/archive-v0/x5d4RtsRYa4it_Ao/lib/python3.12/site-packages/debian/deb822.py:1597:9
     |
1597 |     def get_version(self: _HasVersionFieldProtocol) -> debian.debian_support.Version:
     |         ^^^^^^^^^^^ ------------------------------ Parameter declared here

Found 1 diagnostic
```

As far as I can tell, it's because the name of the `__getitem__` parameter differs.  Maybe it's a bug in the annotations (should `_HasVersionFieldProtocol` have declared `__getitem__`'s parameter to be positional-only?), but nobody expects dunder methods like `__getitem__` to be called with keyword arguments.  Wouldn't it make sense to implicitly consider them to be positional-only?

mypy doesn't complain about this code.  I haven't tried other type checkers.

### Version

0.0.71

**Comment by AlexWaygood** [comment]
Here's a minimal repro of your issue:

```py
from typing import Protocol

class _HasVersionFieldProtocol(Protocol):
    def __getitem__(self, s: str) -> str: ...

class Foo:
    def __getitem__(self, x: str) -> str:
        return ""

x: _HasVersionFieldProtocol = Foo()  # ty error
```

You're right that mypy does not emit an error on this snippet. Every other type checker does, however: ty, pyrefly, pyright, zuban and pycroscope all agree that this is incorrect. Here's a gist you can load into [multiplay](https://github.com/astral-sh/multiplay) to compare them: 54edc25fb9a43420f4b027826652ee5c.

You're right that we certainly _could_ add a heuristic that says all parameters in dunder methods are implicitly positional-only (even though that's not true at runtime). But this is not what mypy is doing here. Mypy also fails to catch type errors on code that very clearly fails at runtime, such as this, because it doesn't take account of parameter names for _any_ methods (not just dunder methods):

```py
from typing import Protocol

class NonDunderProtocol(Protocol):
    def method(self, x): ...

class Bar:
    def method(self, y): ...

def test(obj: NonDunderProtocol):
    obj.method(x=42)

test(Bar())
```

Another multiplay gist -- once again, mypy is the only type checker not to report an error here -- fed8ed7b7ac864abb277fb4c127a83a2. There are several bug reports on the mypy issue tracker pointing out that it is generally much too lenient when it comes to parameter names and questions of subtyping/assignability.

I personally don't think we should change anything here, since ty is clearly in line with the consensus. Explicit calls to dunder methods do happen, and it's just not true that the runtime treats dunder parameters any differently to other parameters.

**Comment by AlexWaygood** [comment]
(I improved the diagnostic hint further in https://github.com/astral-sh/ruff/pull/27717, that'll appear in the next release)
